### PR TITLE
Fix anthropic usage and tools

### PR DESCRIPTION
**Pull request by lazyhope**
This PR fixes two issues with the current implementation:
- In `ANTHROPIC_TOOLS` mode when Claude didn't provide a tool invocation, an index error occurs
- `utils.update_total_usage` only works with OpenAI's `ChatCompletion`, so Anthropic's message usage doesn't update

----

| :rocket: This description was created by [Ellipsis](https://www.ellipsis.dev) for commit 7ceb1b5fa28bca972bd4d882c27b57e8b9eee1f3  | 
|--------|

### Summary:
This PR fixes issues related to Anthropic's usage and tools, updates the `update_total_usage` function to work with Anthropic's message usage, and adds a test case for retry error scenario.

**Key points**:
- Fixed index error in `parse_anthropic_tools` function in `function_calls.py` when there are no tool calls.
- Updated `update_total_usage` function in `utils.py` to work with Anthropic's message usage.
- Modified `retry.py` to handle cases when there is no tool invocation.
- Added a test case in `test_simple.py` to test the retry error scenario.

----
Generated with :heart: by [ellipsis.dev](https://www.ellipsis.dev)

### CODE REVIEW

File: instructor/utils.py [hunk 1]
```diff
@@ -93,11 +94,16 @@ async def extract_json_from_stream_async(
...
-        total_usage.completion_tokens += response.usage.completion_tokens or 0
-        total_usage.prompt_tokens += response.usage.prompt_tokens or 0
-        total_usage.total_tokens += response.usage.total_tokens or 0
+    response_usage = getattr(response, "usage", None)
+    if isinstance(response_usage, OpenAIUsage):
+        total_usage.completion_tokens += response_usage.completion_tokens or 0
+        total_usage.prompt_tokens += response_usage.prompt_tokens or 0
+        total_usage.total_tokens += response_usage.total_tokens or 0
         response.usage = total_usage  # Replace each response usage with the total usage
+    elif isinstance(response_usage, AnthropicUsage):
```
Comment (jxnl) [inline instructor/utils.py:103]: will this break if anthropic is not installed

Comment (lazyhope) [reply instructor/utils.py:103]: You are right, this can be avoided using try import, but models of other providers will still have undefined behavior.

**Comment by lazyhope** [comment]
Also the `RetryError` exception handling is not working in the `retry_async` function: `'RetryError' object has no attribute 'n_attempts'`, and I suppose `raise InstructorRetryException` within the try block appears to be a mistake. @jxnl

File: instructor/function_calls.py [hunk 2]
```diff
@@ -119,9 +119,12 @@ def parse_anthropic_tools(
         validation_context: Optional[Dict[str, Any]] = None,
         strict: Optional[bool] = None,
     ) -> BaseModel:
-        tool_call = [c.input for c in completion.content if c.type == "tool_use"][0]
+        tool_calls = [c.input for c in completion.content if c.type == "tool_use"]
+        tool_call = tool_calls[0] if tool_calls else {}
```
Comment (jxnl) [inline instructor/function_calls.py:123]: we should throw an error there right?

like assert len(tool_calls) == 1, "Please call exactly one tool"

File: instructor/utils.py [hunk 3]
```diff
@@ -93,11 +93,24 @@ async def extract_json_from_stream_async(
...
+
+    # Anthropic usage
+    try:
+        from anthropic.types import Usage as AnthropicUsage
+
+        if isinstance(response_usage, AnthropicUsage):
+            total_usage.input_tokens += response_usage.input_tokens or 0
+            total_usage.output_tokens += response_usage.output_tokens or 0
+            response.usage = total_usage
+    except ImportError:
```
Comment (jxnl) [inline instructor/utils.py:111]: log the exception here.

**Review by jxnl — CHANGES_REQUESTED** [review]
this looks great, left 2 comments

File: instructor/function_calls.py (hunk 2 above)
Comment (lazyhope) [reply instructor/function_calls.py:123]: I left the error indication here:
https://github.com/jxnl/instructor/blob/230276e7067ec232448ed7394c95d1b697a87173/instructor/retry.py#L84
By passing an empty object, the language model will also understand what fields are missing

File: instructor/utils.py (hunk 3 above)
Comment (lazyhope) [reply instructor/utils.py:111]: I suppose we should log outside the exception handling part to indicate that no compatible usage was found? (The usage is neither OpenAI's `ChatCompletionUsage` nor `Anthropic` Usage)

File: instructor/function_calls.py (hunk 2 above)
Comment (lazyhope) [reply instructor/function_calls.py:123]: But it may miss models whose fields all has a default value, let me change that

**Review by jxnl — APPROVED** [review]

**Pull request merged**